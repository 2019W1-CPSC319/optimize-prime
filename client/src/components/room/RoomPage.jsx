import React from 'react';
import moment from 'moment';
import Swal from 'sweetalert2';
import axios from 'axios';
import RoomDialog from './RoomDialog';
import CheckCircleIcon from '@material-ui/icons/CheckCircle';
import RoomRoundedIcon from '@material-ui/icons/RoomRounded';
import InsertInvitationRoundedIcon from '@material-ui/icons/InsertInvitationRounded';
import AddRoundedIcon from '@material-ui/icons/AddRounded';
import { green } from '@material-ui/core/colors';
import {
    Box,
    Table,
    TableHead,
    TableRow,
    TableCell,
    TableBody,
    Paper,
    Fab,
    SnackbarContent,
    Snackbar,
    Slide,
    Button,
} from '@material-ui/core';
import { random } from 'node-forge';

const swalWithBootstrapButtons = Swal.mixin({
    customClass: {
        confirmButton: 'btn btn-success',
        cancelButton: 'btn btn-danger'
    },
    buttonsStyling: true
})

export default class RoomPage extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            rooms: [
                { id: 0, name: '100A', seats: 10 },
                { id: 1, name: '101A', seats: 6 },
                { id: 2, name: '102A', seats: 3 },
                { id: 3, name: '200A', seats: 4 },
                { id: 4, name: '200B', seats: 4 },
                { id: 5, name: '202A', seats: 5 },
                { id: 6, name: '300A', seats: 10 },
                { id: 7, name: '301A', seats: 20 },
                { id: 8, name: '302A', seats: 5 },
                { id: 9, name: '302B', seats: 10 },
            ],
            roomOpen: false,
            onSuccess: false,
        };
    }

    handleOpenAddRoom = () => {
        this.setState({ roomOpen: true });
    }

    handleCloseAddRoom = () => {
        swalWithBootstrapButtons.fire({
            title: 'Are you sure?',
            text: "You won't be able to revert this!",
            type: 'warning',
            showCancelButton: true,
            confirmButtonText: 'PROCEED',
            cancelButtonText: 'CANCEL',
            reverseButtons: true
        }).then((result) => {
            if (result.value) {
                swalWithBootstrapButtons.fire(
                    'Cancelled',
                    'Your progress has not been saved!',
                    'error'
                )
                this.setState({ roomOpen: false });
            } else if (
                result.dismiss === Swal.DismissReason.cancel
            ) {
                this.setState({ roomOpen: false });
            }
        })
        this.setState({ roomOpen: false });
    }

    handleSaveAddRoom = (event) => {
        try {
            event.preventDefault();
            let data = {
                name: event.target[0].value,
                seats: event.target[1].value,
            }
            console.log(data)
            axios.post(`/schedule/room`, data).then(res => {
                console.log(res);
                Swal.fire(
                    'Success',
                    'Successfully added new user',
                    'success'
                );
            }).catch(error => {
                console.error(error);
            });
        }
        catch (err) {
            console.error(JSON.stringify(err));
        }
        this.setState({ roomOpen: false });
        this.setState({ onSuccess: true });
    }

    handleDeleteRoom = (id) => {
        console.log(id);
        try {
            axios.put(`/schedule/room/${id}`).then(res => {
                console.log(res);
                // this.setState({
                //     rooms: res.data
                // });
            }).catch(error => {
                console.error(error);
            });
        }
        catch (err) {
            console.error(JSON.stringify(err));
        }
    }

    showSnackbarOnSuccess = () => {
        return (
            <Snackbar
                anchorOrigin={{
                    vertical: 'bottom',
                    horizontal: 'left',
                }}
                open={this.state.onSuccess}
                autoHideDuration={3000}
                onClose={() => this.setState({ onSuccess: false })}
            >
                <SnackbarContent
                    style={{ backgroundColor: green[600], }}
                    aria-describedby="client-snackbar"
                    message={
                        <Box component='span' id="client-snackbar" style={{
                            display: 'flex',
                            alignItems: 'center',
                        }}>
                            <CheckCircleIcon style={{ fontSize: 20, marginRight: '10px' }} />This is a success message!
                        </Box>
                    }
                />
            </Snackbar>
        )
    }

    async componentDidMount() {
        const { pageProps } = this.props;

        // if (!pageProps.user.profile) {
        //     pageProps.actions.findMeetingTimes();
        // }

        try {
            axios.get('/schedule/rooms').then(res => {
                console.log(res.data)
                this.setState({
                    rooms: res.data
                });
            }).catch(error => {
                console.error(error);
            });
        }
        catch (err) {
            console.error(JSON.stringify(err));
        }
    }

    render() {
        return (
            <div>
                <h1 style={{ marginLeft: '30px', fontWeight: 'normal' }}>Room</h1>
                <Paper square>
                    <Table>
                        <TableHead>
                            <TableRow>
                                <TableCell scope="col">Room Name</TableCell>
                                <TableCell scope="col" align="center">Maximum Capacity</TableCell>
                                <TableCell></TableCell>
                            </TableRow>
                        </TableHead>
                        <TableBody>
                            {this.state.rooms.map(
                                room => {
                                    if (room.status === 'A')
                                        return (
                                            <TableRow key={140}>
                                                <TableCell>{room.name}</TableCell>
                                                <TableCell align="center">{room.seats}</TableCell>
                                                <TableCell>
                                                    <Button
                                                        value={room.roomId}
                                                        color="default"
                                                        variant="outlined"
                                                        onClick={() => this.handleDeleteRoom(room.roomId)}
                                                    >Delete</Button>
                                                </TableCell>
                                            </TableRow>
                                        );
                                    else return <TableRow></TableRow>
                                })}
                        </TableBody>
                    </Table>
                </Paper>
                <Fab
                    color="primary"
                    aria-label="add"
                    style={{ position: 'fixed', right: '30px', bottom: '30px', backgroundColor: '#003b9a' }}
                    onClick={this.handleOpenAddRoom}
                >
                    <AddRoundedIcon />
                </Fab>
                {
                    this.state.roomOpen &&
                    <RoomDialog
                        handleCloseAddRoom={this.handleCloseAddRoom}
                        handleSaveAddRoom={this.handleSaveAddRoom}
                        {...this.state}
                    ></RoomDialog>
                }
                {this.showSnackbarOnSuccess()}
            </div >
        );
    }
}
