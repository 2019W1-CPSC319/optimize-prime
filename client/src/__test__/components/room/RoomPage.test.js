import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import RoomPage from '../../../components/room/RoomPage';

describe('Room Page - Render', () => {
  let wrapper;
  const classes = {};
  const actions = {
    getRooms: jest.fn(),
    getOutlookRooms: jest.fn(),
  };

  beforeEach(() => {
    wrapper = shallow(<RoomPage
      classes={classes}
      actions={actions}
      rooms={[]}
    />);
  });

  it('render the component', () => {
    expect(wrapper.length).toEqual(1);
  });

  it('match snapshot', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
