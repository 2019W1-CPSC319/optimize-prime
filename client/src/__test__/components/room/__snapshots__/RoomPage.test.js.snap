// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Room Page - Render match snapshot 1`] = `
<div>
  <h1
    style={
      Object {
        "fontWeight": "normal",
        "marginLeft": "30px",
      }
    }
  >
    Rooms
  </h1>
  <WithStyles(ForwardRef(Paper))
    style={
      Object {
        "margin": "0 30px",
      }
    }
  >
    <WithStyles(ForwardRef(Table))>
      <WithStyles(ForwardRef(TableHead))>
        <WithStyles(ForwardRef(TableRow))>
          <WithStyles(ForwardRef(TableCell))
            scope="col"
          >
            Room Name
          </WithStyles(ForwardRef(TableCell))>
          <WithStyles(ForwardRef(TableCell))
            align="center"
            scope="col"
          >
            Maximum Capacity
          </WithStyles(ForwardRef(TableCell))>
          <WithStyles(ForwardRef(TableCell)) />
        </WithStyles(ForwardRef(TableRow))>
      </WithStyles(ForwardRef(TableHead))>
      <WithStyles(ForwardRef(TableBody)) />
    </WithStyles(ForwardRef(Table))>
  </WithStyles(ForwardRef(Paper))>
  <WithStyles(ForwardRef(Fab))
    aria-label="add"
    color="primary"
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "#003b9a",
        "bottom": "30px",
        "position": "fixed",
        "right": "30px",
      }
    }
  >
    <AddRoundedIcon />
  </WithStyles(ForwardRef(Fab))>
  <WithStyles(RoomDialog)
    actions={
      Object {
        "getOutlookRooms": [MockFunction] {
          "calls": Array [
            Array [],
            Array [],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        },
        "getRooms": [MockFunction] {
          "calls": Array [
            Array [],
            Array [],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        },
      }
    }
    classes={Object {}}
    error={false}
    handleChangeRoomName={[Function]}
    handleChangeRoomSeats={[Function]}
    handleCloseAddRoom={[Function]}
    handleSaveAddRoom={[Function]}
    name=""
    onSuccess={false}
    roomOpen={false}
    rooms={Array []}
    seats={0}
  />
</div>
`;
